src/wallet/wallet.cpp

```diff
@@ -3046,69 +3005,105 @@ bool CWallet::CreateTransactionInternal(
                     return false;
                 }
 
-                nFeeNeeded = coin_selection_params.m_effective_feerate.GetFee(nBytes);
-                if (nFeeRet >= nFeeNeeded) {
-                    // Reduce fee to only the needed amount if possible. This
-                    // prevents potential overpayment in fees if the coins
-                    // selected to meet nFeeNeeded result in a transaction that
-                    // requires less fee than the prior iteration.
-
-                    // If we have no change and a big enough excess fee, then
-                    // try to construct transaction again only without picking
-                    // new inputs. We now know we only need the smaller fee
-                    // (because of reduced tx size) and so we should add a
-                    // change output. Only try this once.
-                    if (nChangePosInOut == -1 && nSubtractFeeFromAmount == 0 && pick_new_inputs) {
-                        unsigned int tx_size_with_change = nBytes + coin_selection_params.change_output_size + 2; // Add 2 as a buffer in case increasing # of outputs changes compact size
-                        CAmount fee_needed_with_change = coin_selection_params.m_effective_feerate.GetFee(tx_size_with_change);
-                        CAmount minimum_value_for_change = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
-                        if (nFeeRet >= fee_needed_with_change + minimum_value_for_change) {
-                            pick_new_inputs = false;
-                            nFeeRet = fee_needed_with_change;
-                            continue;
+                nFeeRet = coin_selection_params.m_effective_feerate.GetFee(nBytes);
+                // Try to reduce change to include necessary fee
+                if (nSubtractFeeFromAmount == 0) {
+                    assert(nChangePosInOut != -1);
+                    std::vector<CTxOut>::iterator change_position = txNew.vout.begin()+nChangePosInOut;
+                    change_position->nValue -= nFeeRet;
+
+                    if (change_position->nValue < 0) {
+                        CAmount orig_fee = nFeeRet;
+                        // Maybe the change value was just enough to cover the fee if we did not have change in this transaction
+                        // Let's try dropping the change, recalculating the fee, and comparing the new fee to change_and_Fee
+                        nChangePosInOut = -1;
+                        txNew.vout.erase(change_position);
+
+                        tx_sizes = CalculateMaximumSignedTxSize(CTransaction(txNew), this, coin_control.fAllowWatchOnly);
+                        nBytes = tx_sizes.first;
+                        nFeeRet = coin_selection_params.m_effective_feerate.GetFee(nBytes);
```

Comment:
In commit "Move output reductions for fee to after coin selection" (2b445b7c1336506d43874a7ee31f85f511cd3114)

Just a code simplifcation suggestion, but it seems like instead of introducing an orig_fee variable and reducing nFeeRet and then reassigning nFeeRet back to orig_fee, it would be simpler to drop orig_fee and just say:

```c++
CAmount required_fee = coin_selection_params.m_effective_feerate.GetFee(nBytes);
if (required_fee < change_and_fee) {
   nFeeRet = required_fee; // or maybe = change_and_fee, see other comment below
   break;
}
```